### PR TITLE
feat: add support for OTel context propagation and harmonized spans (v3.x)

### LINKED ISSUE #1389: OpenTelemetry integration misses the point: we need to propagate the span/context!
The code where the Subscriber reads the span from the attributes just creates the span and then throws it away:
https://github.com/googleapis/nodejs-pubsub/blob/99107930bbed8e29834acb2347f65efc6a39958f/src/subscriber.ts#L517-L527

A good OpenTelemetry customer would then set the span to the current span in the `on("message")` handler, so that any subsequent processing is attached to the remote span.

I'll also file a Google support case.

#### Environment details

  - `@google-cloud/pubsub` version: master / v2.17.0

#### Steps to reproduce

  1. Enable OpenTelemetry in both publisher and subscriber using `{ enableOpenTelemetryTracing: true }` options on topic and subscription.
  2. Perform some processing that creates spans, either manually or (for example) do a Datastore Query or Put.

It can be observed that actions are linked to the subscriber span, as can also be seen in this screenshot (spans take <1ms because nothing is done inside):
Screenshot 2021-08-30 at 16 47 23

Relates to #1078.

### PR DESCRIPTION
**Pull request by feywind**
Work in progress, converting the existing OTel support to use context propagation and more finely grained spans.

Fixes https://github.com/googleapis/nodejs-pubsub/issues/1389

### CODE REVIEW

**Comment by feywind** [comment]
@weyert Hey, glad to see you're interested :D This is an effort to sync across all the supported languages, so it might be a while before it's fully merged/released. When it's a bit further along though, I'd be interested in testing help, since you're one of the people who's more into OTel usage. This is my main project right now, so I'm hoping I can get a preview branch going soon, at the least.

**Comment by weyert** [comment]
Awesome happy to try out a preview version when it is available :)

**Comment by feywind** [comment]
@weyert I'm hoping to have this to a point that others can try it out in a super early preview sort of capacity tomorrow. I mostly want to add a Cloud Trace sample and copy the branch over to the main repo so others can make PRs against it. I'm going to be out for a bit and don't want to hold up playing with the feature :)

**Comment by feywind** [comment]
I think I won't end up having time to get to this after all - the branch is here, if anyone ends up wanting to make any PRs against it. Happy holidays, all :)

https://github.com/googleapis/nodejs-pubsub/tree/otel-2

**Comment by weyert** [comment]
Thank you @feywind! I will give it a try during the holidays :)

**Comment by 0xdevalias** [comment]
Just wanted to check in on this PR and see if there were any new updates for 2023 so far?

**Comment by feywind** [comment]
Nothing yet, but we're trying to get some cycles to dedicate to finishing it.

**Comment by feywind** [comment]
Screenshot 2023-05-08 at 6 12 09 PM

@hongalex ^

**Comment by weyert** [comment]
> Screenshot 2023-05-08 at 6 12 09 PM
> 
> @hongalex ^

Looks great! Any plans to do a prerelease of this? That would make experimenting / testing this a lot easier

**Comment by feywind** [comment]
@weyert Yeah, I can suggest that. There's some sync work going on to make sure all of this stuff is consistent across languages and products.

I know this has been sitting here a while and it might seem like shouting into the void 😹 but I've been able to say in meetings "we have a bunch of users clamouring for this", so the comments are useful :)

**Comment by weyert** [comment]
Any  way I can help with getting this PR merged?

**Comment by feywind** [comment]
> Any way I can help with getting this PR merged?

It's mostly waiting on review at this point. I hate to ask a user to do review work :D but if you have any comments on any of it, please feel free to add them. My colleague @hongalex  will be reviewing it at some point and it might help him go faster.

Since 4.0.0 is going to be almost entirely Node EOS stuff, I can probably back-port this to the 3.x line once it's ready, as well.

File: src/telemetry-tracing.ts
Comment (hongalex) [inline src/telemetry-tracing.ts:285]: we should also be recording how many messages are published in the same batch as an attribute

**Review by hongalex — COMMENTED** [review]
Looks good mostly, just some comments about ways to potentially help with readability

File: samples/openTelemetryTracing.js
Comment (feywind) [reply samples/openTelemetryTracing.js:35]: Yeah, I think the person who originally made this just came up with one? I think maybe it deserves several samples, e.g. one for Cloud Trace, another for console output, etc.

File: src/subscriber.ts
Comment (feywind) [reply src/subscriber.ts:202]: Yeah, I am okay with the renaming comments, I mostly wanted to get away from having "OpenTelemetry" in the code names explicitly.

File: src/subscriber.ts
Comment (feywind) [reply src/subscriber.ts:513]: Yeah, this could be expanded. Basically in the previous iteration of support, you had to set a flag explicitly to enable tracing. In the new one, that flag should only be set if you want the old attributes and measuring methods and such. So the intent here is "unset this flag in the future, and just instantiate a tracing provider so that it uses standard context propagation". The actual wording doesn't need to be that complex, though.

File: src/subscriber.ts
Comment (feywind) [reply src/subscriber.ts:67]: Allow more providers == something besides OpenTelemetry. I was thinking about the switch from OpenCensus. Using data structures and names not specific to OTel seems like a good investment.

File: test/telemetry-tracing.ts
Comment (feywind) [reply test/telemetry-tracing.ts:2]: Same as above :D This is also a rename.

File: src/telemetry-tracing.ts
Comment (feywind) [reply src/telemetry-tracing.ts:127]: Yeah, that seems fine. I don't think there's a particular reason it needs to be one class.

**Comment by feywind** [comment]
I'm going to go ahead and base this off the 3.x branch to get it merged into that, and then I'll up-port it to 4.x.

File: samples/openTelemetryTracing.js
Comment (hongalex) [reply samples/openTelemetryTracing.js:35]: I think we shouldn't do console output and just focus on Cloud Trace as the default sample (since that's also what our open telemetry team's documentation points to).

File: samples/openTelemetryTracing.js
Comment (hongalex) [inline samples/openTelemetryTracing.js:95]: I'm encountering a strange issue where (running this sample and using the cloud trace exporter), the `receive` span seems to get clobbered. Any ideas?

image

File: samples/openTelemetryTracing.js
Comment (feywind) [reply samples/openTelemetryTracing.js:35]: I do think that's the primary thing we want to help people with, 'cause it's probably what they want to do. But I do think we ought to at least e.g. have a Medium article talking about how to output to console, Grafana, etc. Those will be useful to people too.

File: samples/openTelemetryTracing.js
Comment (feywind) [reply samples/openTelemetryTracing.js:95]: This ended up being caused by me missing a bunch of places I needed to follow up on the renames earlier. I did some more renaming on stuff that had the ambiguous names, so hopefully that's cleaner all around.

image

Comment (hongalex) [reply samples/openTelemetryTracing.js:95]: Yeah that fixed the receive span naming. Now I'm having an issue where the spans are still disconnected. Looking at the message directly, it looks like the otel propagation attributes are missing. Any idea what's wrong?

image

File: package.json
Comment (SimenB) [inline package.json:54]: any particular reason these are locked down from `^`?

Comment (feywind) [reply package.json:54]: Yeah, it had to do with Node 12 compatibility. I found that there were some semver issues, in practice, though I'll be revisiting this for the 4.x version of the PR, so I'll probably recheck these for 3.x too.

File: samples/openTelemetryTracing.js
Comment (hongalex) [reply samples/openTelemetryTracing.js:95]: Following up on this, I think there's something wrong I'm doing in my code since the basic sample includes the proper propagation.

Comment (hongalex) [reply samples/openTelemetryTracing.js:95]: Ok I think I found the problem. Calling [`NodeTracerProvider.register()`](https://github.com/googleapis/nodejs-pubsub/blob/b218d5ff999eaf9e87c1cb4154f156679560c312/samples/typescript/openTelemetryTracing.ts#L75) makes this work as intended.

However, this isn't explicitly stated as required in the [Cloud Trace OpenTelemetry Node docs](https://cloud.google.com/trace/docs/setup/nodejs-ot), in [this sample](https://github.com/GoogleCloudPlatform/opentelemetry-operations-js/blob/13997a0256e5124b5dda1c726db3f7c859e0a560/samples/trace/index.js#L48) specifically where it recommends calling `setGlobalTraceProvider(provider)` and does _not_ call `register`.

I'm not entirely sure what the difference is (since if you do both `register` and `setGlobalTraceProvider`), you get the error `Error: @opentelemetry/api: Attempted duplicate registration of API: trace` which seems to indicate these methods are doing the same thing. If `setGlobalTraceProvider` is not enough, we should file an issue to fix that sample.

Comment (hongalex) [reply samples/openTelemetryTracing.js:95]: So using `setGlobalTraceProvider` isn't enough since `register` does additional things with the propagator. The Cloud Trace sample doesn't specify this since they don't use context propagation. As long as people are following our sample, it should work.

**Review by hongalex — DISMISSED** [review]

**Comment by weyert** [comment]
Can't wait for this one to get merged. I have run it locally and works well. I haven't been able to run it in the wild as we had some maddening memory issues but that might have been better

I assume it adheres to the messaging outline of opentelemetry? I have to admit I don't fully understand that one

**Comment by feywind** [comment]
> Can't wait for this one to get merged.

You and I both :)

> I have run it locally and works well. I haven't been able to run it in the wild as we had some maddening memory issues but that might have been better

Memory issue caused by the OTel PR or something unrelated?

> I assume it adheres to the messaging outline of opentelemetry? I have to admit I don't fully understand that one

Such as it is, yeah. The inner workings of the GCP Pub/Sub libraries are represented in the span layout, but I don't think that's particularly specified by OTel. It's our intent to support the standard span attributes and W3C trace propagation attribute.

TBD if the span structure will be identical across client libraries, like C++ has a different thing going that's based on their method call structure.

Given that the 3.x release machinery is just about going, I'm planning to bring this PR up to the main branch at some point and then back-port the finished one to 3.x.

**Comment by weyert** [comment]
> > I have run it locally and works well. I haven't been able to run it in the wild as we had some maddening memory issues but that might have been better
> 
> Memory issue caused by the OTel PR or something unrelated?

Oh no, work code related issue. Nothing related to this PR sorry for confusing you.

 > > I assume it adheres to the messaging outline of opentelemetry? I have to admit I don't fully understand that one
> 
> Such as it is, yeah. The inner workings of the GCP Pub/Sub libraries are represented in the span layout, but I don't think that's particularly specified by OTel. It's our intent to support the standard span attributes and W3C trace propagation attribute.
> 
> TBD if the span structure will be identical across client libraries, like C++ has a different thing going that's based on their method call structure.

Yeah, I understood their is this messaging semantic convention they are working on, like https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-spans.md

Makes sense the span structure is the same structure between libraries :))
 
> Given that the 3.x release machinery is just about going, I'm planning to bring this PR up to the main branch at some point and then back-port the finished one to 3.x.

Makes sense but to be honest I can also fully understand it's a v4 thing

**Comment by feywind** [comment]
Will re-open this later when the 4.x one is done.

**Comment by feywind** [comment]
I didn't plan it this way 😹 but it's past the year support window for 3.x, and a lot of changes have happened since OTel started, so I won't be doing this back-porting work.

**Pull request closed without merging**
